core/src/main/scala/com/evolutiongaming/catshelper/CountLatch.scala

```diff
@@ -0,0 +1,83 @@
+package com.evolutiongaming.catshelper
+
+import cats.effect.{Async, Concurrent, Deferred, Ref}
+import cats.syntax.all._
+
+/**
+  * Analog of [[cats.effect.std.CountDownLatch]] that supports increases of latches after creation via method [[CountLatch.acquire]]
+  *
+  * Example:
+  * {{{
+  *    import cats.effect.IO
+  *    for {
+  *       latch <- CountLatch[IO](1)
+  *       fiber <- latch.await.start
+  *       _ <- latch.acquire
+  *       _ <- latch.release
+  *       _ <- latch.release
+  *       _ <- fiber.joinWithNever
+  *    } yield {}
+  * }}}
+  *
+  * @tparam F effect type, expected to be [[cats.effect.IO]]
+  */
+sealed trait CountLatch[F[_]] {
+
+  /** Increase latches by one */
+  def acquire: F[Unit]
+
+  /** Decrease latches by one */
+  def release: F[Unit]
+
+  /** Semantically blocks fiber while latches more than zero */
+  def await: F[Unit]
+}
+
+object CountLatch {
+
+  def apply[F[_]: Async](n: Int = 0): F[CountLatch[F]] =
+    for {
+      state <- if (n > 0) State.awaiting(n) else State.Done[F]().pure[F]
+      state <- Ref.of[F, State[F]](state)
+    } yield
+      new CountLatch[F] {
+
+        override def acquire: F[Unit] =
+          state.update {
```

Comment:
in case you want to remove `Deferred.unsafe`
```scala
        override def acquire: F[Unit] = {
          0.tailRecM { counter =>
            state
              .access
              .flatMap { case (state, set) =>
                for {
                  state <- state match {
                    case State.Done() =>
                      Deferred[F, Unit].map { deferred =>
                        State.Awaiting(1, deferred)
                      }

                    case State.Awaiting(n, await) =>
                      State.Awaiting(n + 1, await).pure[F]
                  }
                  result <- set(state)
                } yield {
                  if (result) {
                    ().asRight[Int]
                  } else {
                    (counter + 1).asLeft[()]
                  }
                }
              }
          }
        }

```

Reply:
Nice idea!

Reply:
done

Reply:
only downside I can mention - instead of CAS operation on AtomicRef the proposal will use loop under the hood

Reply:
>  instead of CAS operation 

afaik, it's still CAS operation under the hood.